pkg/cosign/tuf/client.go

```diff
@@ -311,27 +301,41 @@ func NewFromEnv(ctx context.Context) (*TUF, error) {
 	}
 
 	// Initializes a new TUF object from the local cache or defaults.
-	return initializeTUF(ctx, embed, mirror, nil, false)
+	return initializeTUF(ctx, mirror, nil, GetEmbedded(), false)
 }
 
 func Initialize(ctx context.Context, mirror string, root []byte) error {
 	// Initialize the client. Force an update.
-	t, err := initializeTUF(ctx, false, mirror, root, true)
+	t, err := initializeTUF(ctx, mirror, root, GetEmbedded(), true)
 	if err != nil {
 		return err
 	}
 	t.Close()
 
-	// Store the remote for later.
-	remoteInfo := &remoteCache{Mirror: mirror}
-	b, err := json.Marshal(remoteInfo)
+	// Store the remote for later if we are caching.
+	if !noCache() {
+		remoteInfo := &remoteCache{Mirror: mirror}
+		b, err := json.Marshal(remoteInfo)
+		if err != nil {
+			return err
+		}
+		if err := os.WriteFile(cachedRemote(rootCacheDir()), b, 0600); err != nil {
+			return fmt.Errorf("storing remote: %w", err)
+		}
+	}
+	return nil
+}
+
+// Checks if the testTarget matches the valid target file metadata.
+func isValidTarget(testTarget []byte, validMeta data.TargetFileMeta) bool {
+	localMeta, err := util.GenerateTargetFileMeta(bytes.NewReader(testTarget))
 	if err != nil {
-		return err
+		return false
 	}
-	if err := os.WriteFile(cachedRemote(rootCacheDir()), b, 0600); err != nil {
-		return fmt.Errorf("storing remote: %w", err)
+	if err := util.TargetFileMetaEqual(localMeta, validMeta); err != nil {
```

Comment:
To also double check - This would catch if the embedded targets are modified for some reason, correct? So to verify the e2e flow:
* Cosign requests the target metadata (targets.json) from the remote
* The metadata's sig is verified using the target role from the root (which has already been updated and verified)
* For each target in targets.json, we check if the target needs to be downloaded or is present in the embedded/local cache
* For each target, we compare its hash to the hash from the metadata (this step)

Reply:
Yes! Mostly correct, except slight details in Target retrieval/download from https://github.com/sigstore/cosign/blob/e7bcb6961178788ceb5a2e63e33388a188b574c9/pkg/cosign/tuf/client.go#L464

* Cosign requests the target metadata (targets.json) from the remote
* The metadata's sig is verified using the target role from the root (which has already been updated and verified)
* For each target in targets.json, we check (1) if we have the target in our local cache, if it's hash is valid, do nothing. (2) if the target is present in the embedded/local cache, we compare its hash from the metadata and copy it to the cache for ease of retrieval or (3) it must need to be downloaded (and the hash is cerified in the go-tuf call). 
* On get, we retrieve the target compare its hash to the hash from the metadata (this step)